### ISSUE TITLE
Usage in NodeJS environment

### ISSUE THREAD

**Issue by Lukas-Kullmann** [jaydenseric/apollo-upload-client#129]
I try to send a request from a NodeJS server to the GraphQL server. Unfortunately, Node does neither support `File`s nor `Blob`s. So I figured, I'd use the `ReactNativeFile`, but it told me, that `FormData` is undefined.

So I used [`form-data`](https://www.npmjs.com/package/form-data) as a Polyfill. The problem is, it can't handle `ReactNativeFile`s (somwhere down the `FormData.append` it throws `source.on is not a function` where `source` is the `ReactNativeFile`).

Am I missing something? And if not, is it possible to add support for `Buffer`, readable streams or something that allows file uploads with Node?

**Comment by jaydenseric** [comment]
What are you trying to do? You can make a [GraphQL multipart request](https://github.com/jaydenseric/graphql-multipart-request-spec) in Node.js like this: https://github.com/jaydenseric/graphql-upload/blob/8d0f3e5ad144ab15eb7449e53322627eaed1d541/src/test.mjs#L77

**Comment by jaydenseric** [comment]
Well, actually that example is a little simplistic as it doesn't include the operation `query`, as a full GraphQL server is not being tested. But just follow the spec, using `FormData` and `fetch`.

**Comment by Lukas-Kullmann** [comment]
Sorry if this was not clear.

I try to send a GraphQL mutation from a Node environment:

```js
const client = new ApolloClient({link: createUploadLink()});

await client.mutate({
  mutation: gql`
    mutation uploadFile($file: Upload!) {
      # something
    }
  `,
  variables: {
    file: /* what to but here? */
  }
});
```

And my question is what to put as a file to be able to send the mutation.

**Comment by jaydenseric** [comment]
Why would you want to use Apollo Client?

**Comment by Lukas-Kullmann** [comment]
Well, I do not want to configure the request manually. The ApolloClient API is very convenient. And for non-file requests it works like a charm.

**Comment by jaydenseric** [comment]
For your use, all Apollo Client is doing is making a `fetch` request, something you could do yourself without the overhead. On a server, and particularly with mutations, you probably don't want any caching and such to prevent contamination between your server requests. The less work your server does (and the less dependencies to manage) the better.

Regardless, Node.js does not support the `File` and `FileList` API so you basically can't server test file upload mutations via Apollo Client. I spent a lot of time in the past trying. Frankly it's stupid Node.js doesn't support APIs like that, `FormData` and `fetch` out of the box; [Deno](https://github.com/denoland/deno) will.

**Comment by Lukas-Kullmann** [comment]
Yeah, you are right. I will make the requests manually then.

Thanks for the fast reply!

**Comment by jaydenseric** [comment]
To hypothetically support file uploads from a Node.js environment (without `File` and `FileList`) we would need to add support for detecting file streams (and only _file_ streams, not just any stream) to [`extract-files`](https://github.com/jaydenseric/extract-files). [This has briefly been discussed before](https://github.com/jaydenseric/apollo-upload-client/pull/79#issuecomment-380669915) for remote schema merging.

**Comment by andreisocaciu** [comment]
@jaydenseric When can we expect this to work with the newly released [extract-files@6.0.0](https://github.com/jaydenseric/extract-files/pull/11)?
